File: LeetCodeProblems/Problems/LeetCode190ReverseBits.cs
namespace LeetCodeProblems.Problems;

public class LeetCode190ReverseBits
{
    public uint reverseBits(uint n)
    {
        Stack<bool> binary = [];

        for (int i = 0; i < 32; i++)
        {
            binary.Push(n % 2 == 1);
            n /= 2;
        }

        uint result = 0;
        uint coefficient = 0;
        for (int i = 0; i < 32; i++)
        {
            coefficient=(uint)(binary.Pop()? 1 : 0);
            if(coefficient == 0)
                continue;
            result += coefficient*(uint)Math.Pow(2, i);
        }

        return result;
    }
}

File: LeetCodeProblemsGo/LeetCode190ReverseBits/LeetCode190ReverseBits.go
package LeetCode190ReverseBits

func reverseBits(n int) int {
	first, second := 0, 0
	for i := 0; i < 16; i++ {
		first = 0
		if (n & (1 << i)) > 0 {
			first = 1
		}
		second = 0
		if (n & (1 << (31 - i))) > 0 {
			second = 1
		}
		first = first ^ second
		second = first ^ second
		first = first ^ second
		n = (n & ^(1 << i)) | (first << i)
		n = (n & ^(1 << (31 - i))) | (second << (31 - i))
	}
	return n

}
